sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -907,29 +907,64 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
   }
 
   /**
-   * Add an [[Aggregate]] or [[GroupingSets]] to a logical plan.
+   * Add an [[Aggregate]] to a logical plan.
    */
   private def withAggregationClause(
       ctx: AggregationClauseContext,
       selectExpressions: Seq[NamedExpression],
       query: LogicalPlan): LogicalPlan = withOrigin(ctx) {
-    val groupByExpressions = expressionList(ctx.groupingExpressions)
-
-    if (ctx.GROUPING != null) {
-      // GROUP BY .... GROUPING SETS (...)
-      val selectedGroupByExprs =
-        ctx.groupingSet.asScala.map(_.expression.asScala.map(e => expression(e)).toSeq)
-      GroupingSets(selectedGroupByExprs.toSeq, groupByExpressions, query, selectExpressions)
-    } else {
-      // GROUP BY .... (WITH CUBE | WITH ROLLUP)?
-      val mappedGroupByExpressions = if (ctx.CUBE != null) {
-        Seq(Cube(groupByExpressions))
-      } else if (ctx.ROLLUP != null) {
-        Seq(Rollup(groupByExpressions))
+    if (ctx.groupingExpressionsWithGroupingAnalytics.isEmpty) {
+      val groupByExpressions = expressionList(ctx.groupingExpressions)
+      if (ctx.GROUPING != null) {
+        // GROUP BY .... GROUPING SETS (...)
+        val selectedGroupByExprs =
+          ctx.groupingSet.asScala.map(_.expression.asScala.map(e => expression(e)).toSeq)
```

Comment:
Seems we don't distinguish `(a)` and `a`, then how do we handle `GROUPING SETS(a, b, c)` and `GROUPING SETS((a), (b), (c))`?

According to the document, `GROUPING SETS(a, b, c)` is `GROUPING SETS((a, b), (a, c), (b, c), (a), (b), (c))`.

Reply:
or is my understanding totally wrong? `a` is the same as `(a)`? Then we should update the document in https://github.com/apache/spark/pull/30212/files#diff-e895c94a06560e4db8e99be99cc7864c065c2302402dcd158cd825f3b09f878fR63

Reply:
> Seems we don't distinguish `(a)` and `a`, then how do we handle `GROUPING SETS(a, b, c)` and `GROUPING SETS((a), (b), (c))`?
> 
> According to the document, `GROUPING SETS(a, b, c)` is `GROUPING SETS((a, b), (a, c), (b, c), (a), (b), (c))`.

`GROUPING SETS(a, b, c)`  is same as  `GROUPING SETS((a), (b), (c))`?

Reply:
here the group set  depending on the parser logic right?
https://github.com/apache/spark/blob/8945e7c52f8d09282f6847c19bd926ac4c20e11a/sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4#L604-L611

According to the definition, `a` can be a grouping_set, `(a)` can be a grouping_set, `(a, b)` can be a grouping_set.

Reply:
For SQL
```
select id, city, car_model, sum(quantity)
from dealer group by grouping sets((id), (city), (car_model))
UNION all
select id, city, car_model, sum(quantity)
from dealer group by grouping sets(id, city, car_model)
order by id, city, car_model
```
Result in PGSQL
![image](https://user-images.githubusercontent.com/46485123/112615257-3b203000-8e5d-11eb-958e-fb5196c9592c.png)